toolkit/offchain/src/register.rs

```diff
@@ -1,18 +1,174 @@
+	TransactionContext,
+};
+use crate::{
+	await_tx::{AwaitTx, FixedDelayRetries},
+	plutus_script::PlutusScript,
+	OffchainError,
+};
+use anyhow::anyhow;
 use cardano_serialization_lib::{
-	BigNum, DataCost, ExUnits, JsError, MinOutputAdaCalculator, Transaction, TransactionBuilder,
-	TransactionOutputBuilder, TxInputsBuilder,
+	BigNum, DataCost, ExUnits, JsError, MinOutputAdaCalculator, PlutusData, Transaction,
+	TransactionBuilder, TransactionOutputBuilder, TxInputsBuilder,
+};
+use ogmios_client::{
+	query_ledger_state::{QueryLedgerState, QueryUtxoByUtxoId},
+	query_network::QueryNetwork,
+	transactions::Transactions,
+	types::OgmiosTx,
+	types::OgmiosUtxo,
+};
+use partner_chains_plutus_data::registered_candidates::{
+	candidate_registration_to_plutus_data, RegisterValidatorDatum,
 };
-use ogmios_client::types::OgmiosUtxo;
-use partner_chains_plutus_data::registered_candidates::block_producer_registration_to_plutus_data;
-use sidechain_domain::BlockProducerRegistration;
+use sidechain_domain::*;
+
+pub trait Register {
+	#[allow(async_fn_in_trait)]
+	async fn register(
+		&self,
+		genesis_utxo: UtxoId,
+		candidate_registration: &CandidateRegistration,
+		payment_signing_key: MainchainPrivateKey,
+	) -> Result<Option<OgmiosTx>, OffchainError>;
+}
+
+impl<T> Register for T
+where
+	T: QueryLedgerState + Transactions + QueryNetwork + QueryUtxoByUtxoId,
+{
+	async fn register(
+		&self,
+		genesis_utxo: UtxoId,
+		candidate_registration: &CandidateRegistration,
+		payment_signing_key: MainchainPrivateKey,
+	) -> Result<Option<OgmiosTx>, OffchainError> {
+		run_register(
+			genesis_utxo,
+			candidate_registration,
+			payment_signing_key,
+			self,
+			FixedDelayRetries::two_minutes(),
+		)
+		.await
+		.map_err(|e| OffchainError::InternalError(e.to_string()))
+	}
+}
+
+pub async fn run_register<
+	C: QueryLedgerState + QueryNetwork + QueryUtxoByUtxoId + Transactions,
+	A: AwaitTx,
+>(
+	genesis_utxo: UtxoId,
+	candidate_registration: &CandidateRegistration,
+	payment_signing_key: MainchainPrivateKey,
+	ogmios_client: &C,
+	await_tx: A,
+) -> anyhow::Result<Option<OgmiosTx>> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key.0, ogmios_client).await?;
+	let validator = crate::scripts_data::registered_candidates_scripts(genesis_utxo)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let registration_utxo = ctx
+		.payment_key_utxos
+		.iter()
+		.find(|u| u.to_domain() == candidate_registration.registration_utxo)
+		.ok_or(anyhow!("registration utxo not found at payment address"))?;
+	let all_registration_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
+	let own_registrations = get_own_registrations(
+		candidate_registration.own_pkh,
+		candidate_registration.stake_ownership.pub_key.clone(),
+		&all_registration_utxos,
+	);
+	let own_registration_utxos = own_registrations.iter().map(|r| r.0.clone()).collect::<Vec<_>>();
```

Comment:
cosmetic: this could go lower, because now it breaks into checking if there exists registration with given keys